pkg/apiserver/registry/generic/key.go

```diff
@@ -12,63 +12,69 @@ import (
 )
 
 type Key struct {
-	Group     string
-	Resource  string
-	Namespace string
-	Name      string
+	Group     string `json:"group,omitempty"`
+	Resource  string `json:"resource"`
+	Namespace string `json:"namespace,omitempty"`
+	Name      string `json:"name,omitempty"`
 }
 
-func ParseKey(key string) (*Key, error) {
-	// /<group>/<resource>[/namespaces/<namespace>][/<name>]
-	parts := strings.Split(key, "/")
-	if len(parts) < 3 {
-		return nil, fmt.Errorf("invalid key (expecting at least 2 parts): %s", key)
-	}
-
-	if parts[0] != "" {
-		return nil, fmt.Errorf("invalid key (expecting leading slash): %s", key)
-	}
-
-	k := &Key{
-		Group:    parts[1],
-		Resource: parts[2],
-	}
-
-	if len(parts) == 3 {
-		return k, nil
-	}
+// ParseKey parses a key string into a Key.
+// Format: [/group/<group>]/resource/<resource>[/namespace/<namespace>][/name/<name>]
+func ParseKey(raw string) (*Key, error) {
+	parts := strings.Split(raw, "/")
+	key := &Key{}
 
-	if parts[3] != "namespaces" {
-		k.Name = parts[3]
-		return k, nil
+	// Skip the first empty string
+	if parts[0] == "" {
+		parts = parts[1:]
 	}
 
-	if len(parts) < 5 {
-		return nil, fmt.Errorf("invalid key (expecting namespace after 'namespaces'): %s", key)
+	for i := 0; i < len(parts); i += 2 {
+		k := parts[i]
+		if i+1 >= len(parts) {
+			return nil, fmt.Errorf("invalid key: %s", raw)
+		}
+		v := parts[i+1]
+		switch k {
+		case "group":
+			key.Group = v
+		case "resource":
+			key.Resource = v
+		case "namespace":
+			key.Namespace = v
+		case "name":
+			key.Name = v
+		default:
+			return nil, fmt.Errorf("invalid key name: %s", key)
+		}
 	}
 
-	k.Namespace = parts[4]
-
-	if len(parts) == 5 {
-		return k, nil
+	if len(key.Resource) == 0 {
+		return nil, fmt.Errorf("missing resource: %s", raw)
 	}
 
-	k.Name = parts[5]
-
-	return k, nil
+	return key, nil
 }
 
+// String returns the string representation of the Key.
 func (k *Key) String() string {
-	s := "/" + k.Group + "/" + k.Resource
+	s := ""
```

Comment:
i never know if string contact vs strings.Buidler is the right approach -- do you know?

likely does not matter -- mostly for my curiosity

Reply:
it sounds like it depends on the length of the string. i'll benchmark

Reply:
```
BenchmarkKey_StringBuilder-32    9452661	       130.8 ns/op	     176 B/op	       4 allocs/op
BenchmarkKey_String-32    	 5975756	       196.3 ns/op	     232 B/op	       4 allocs/op
```

Reply:
updated in https://github.com/grafana/grafana/pull/90014/commits/7a0a3d72489d94efcddaec62611b9947b7b571d4